File: competence-database/src/main/scala/uzuzjmd/competence/persistence/dao/Competence.scala
package uzuzjmd.competence.persistence.dao

import uzuzjmd.competence.config.Logging
import uzuzjmd.competence.persistence.abstractlayer.{OntResult, CompOntologyManager, CompOntologyAccess}
import uzuzjmd.competence.persistence.ontology.CompObjectProperties
import uzuzjmd.competence.persistence.ontology.CompOntClass
import com.hp.hpl.jena.rdf.model.Property
import com.hp.hpl.jena.rdf.model.Statement
import com.hp.hpl.jena.ontology.OntClass
import uzuzjmd.competence.persistence.owl.{CompOntologyAccessScala, CompOntologyManagerJenaImpl}
import scala.collection.JavaConverters._
import uzuzjmd.competence.exceptions.{IndividualNotFoundException, NoRecursiveSubClassException, DataFieldNotInitializedException}

class Competence(compManager: CompOntologyManager,
                 identifierLocal: String,
                 val definition: String = null,
                 val compulsory: java.lang.Boolean = null) extends CompetenceOntologySingletonDao(compManager, CompOntClass.Competence, identifierLocal) with Logging {


  def COMPULSORY = "compulsory"

  def getCompulsory() = getDataField(COMPULSORY)

  @Override
  protected def persistMore() {
    val competenceRoot = new CompetenceInstance(comp)
    val ontClass = getOntClass;
    var rootOntClass = competenceRoot.getOntClass;
    if (rootOntClass == null) {
      rootOntClass = competenceRoot.persist.getOntclass
    }
    ontClass.addSuperClass(rootOntClass)
    if (definition != null) {
      addDataField(DEFINITION, definition)
    } else {
      addDataField(DEFINITION, identifierLocal)
    }
    if (compulsory == null) {
      addDataField(COMPULSORY, new java.lang.Boolean(false))
    } else {
      addDataField(COMPULSORY, compulsory)
    }

    addCourseContext(new CourseContext(compManager, "university"))
  }




  @Override
  def getFullDao(): Competence = {
   /* if (this.identifierTop.equals("ComptenceIndividual")) {
      return this;
    }*/

    val individual = getIndividual
    if (individual == null) {
      logger.error("Individual not found for id: "+ identifierLocal)
      throw new IndividualNotFoundException();
    }
    val definition2 = getDefinition
    try {
      getDataFieldBoolean(COMPULSORY)
    } catch {
      case e:DataFieldNotInitializedException => new Competence(compManager, identifier, definition2, false);
    }
    return new Competence(compManager, identifier, definition2, getDataFieldBoolean(COMPULSORY))
  }

  def addSuggestedCompetenceRequirement(competence: Competence) {
    createEdgeWith(competence, CompObjectProperties.SuggestedCompetencePrerequisiteOf)
  }

  def getSuggestedCompetenceRequirements(): List[Competence] = {
    return getAssociatedSingletonDaosAsDomain(CompObjectProperties.SuggestedCompetencePrerequisiteOf, classOf[Competence])
  }

  def addRequiredCompetence(competence: Competence) {
    deleteEdgeWith(competence, CompObjectProperties.NotPrerequisiteOf)
    createEdgeWith(competence, CompObjectProperties.PrerequisiteOf)
  }

  def addNotRequiredCompetence(competence: Competence) {
    deleteEdgeWith(competence, CompObjectProperties.PrerequisiteOf)
    createEdgeWith(competence, CompObjectProperties.NotPrerequisiteOf)
  }

  def getCatchwords(): List[Catchword] = {
    val result = getAssociatedSingletonDaosAsDomain(CompObjectProperties.CatchwordOf, classOf[Catchword])
    return result
  }

  def getRequiredCompetences(): List[Competence] = {
    return getAssociatedSingletonDaosAsDomain(CompObjectProperties.PrerequisiteOf, classOf[Competence])
  }

  def getRequiredCompetencesAsArray(): Array[String] = {
    return getRequiredCompetences().map(x => x.getDataField(x.DEFINITION)).toArray;
  }

  def isLinkedAsRequired(): Boolean = {
    return !(getRequiredCompetences.isEmpty && getAssociatedSingletonDaosAsRange(CompObjectProperties.PrerequisiteOf, classOf[Competence]).isEmpty);
  }

  def isAllowed(user: User): Boolean = {
    val prerequisites = getAssociatedSingletonDaosAsDomain(CompObjectProperties.PrerequisiteOf, classOf[Competence])
    if (prerequisites.isEmpty) {
      return true
    } else {
      return prerequisites.forall(x => x.hasEdge(user, CompObjectProperties.UserHasPerformed) && x.isAllowed(user))
    }
  }

  def addSuperCompetence(superCompetence: Competence): Competence = {
    persistManualCascades(false).getOntclass().addSuperClass(superCompetence.persistManualCascades(true).getOntclass())
    persistManualCascades(false)
    return this
  }

  def removeSuperCompetence(superCompetence: Competence): Competence = {
    persistManualCascades(false).getOntclass().removeSuperClass(superCompetence.persistManualCascades(true).getOntclass())
    persistManualCascades(false)
    return this
  }

  def toStrinz(): String = {
    return getDefinition()
  }

  def getAssessment(user: User): SelfAssessment = {
    val result1 = getAssociatedStandardDaosAsDomain(CompObjectProperties.AssessmentOfCompetence, classOf[SelfAssessment])
    val result = result1.filter(x => x.hasEdge(CompObjectProperties.AssessmentOfUser, user))
    if (result.isEmpty) {
      return new SelfAssessment(compManager, CompOntologyAccessScala.createIdentifierForAssessment(user, this), this, user, 0, false)
    }
    return result.head
  }

  def getShortestPathToSubCompetence(subCompetence: Competence): java.util.List[Competence] = {
    return compManager.getUtil().getShortestSubClassPath(subCompetence.toOntClass, toOntClass).asScala.map(x => new Competence(compManager, x, null, null)).asJava
  }

  def getCatchwordsAsJava(): java.util.List[Catchword] = {
    return getCatchwords.asJava
  }

  def addCatchword(dao: Catchword) {
    dao.persistManualCascades(true)
    dao.createEdgeWith(CompObjectProperties.CatchwordOf, this)
  }

  def addLearningTemplate(learningTemplate: LearningProjectTemplate) {
    createEdgeWith(learningTemplate, CompObjectProperties.LearningProjectTemplateOf)
  }

  def getOperators(): List[Operator] = {
    return getAssociatedSingletonDaosAsDomain(CompObjectProperties.OperatorOf, classOf[Operator]);
  }

  def addCourseContext(course: CourseContext) {
    createEdgeWith(course, CompObjectProperties.CourseContextOf)
    addSuperCompetencesToCourse(this, course)
  }

  private def addSuperCompetencesToCourse(competence: Competence, course: CourseContext) {
    if (competence.hasSuperClass) {
      val superCompetences = competence.listSuperClasses(classOf[Competence]).filterNot { x => x.identifier.equals("Competence") }.filterNot { x => x.identifier.equals("Resource") }
      superCompetences.foreach { x => x.createEdgeWith(course, CompObjectProperties.CourseContextOf) }
      superCompetences.foreach { x => addSuperCompetencesToCourse(x, course) }
    }
  }

  override def canEqual(other: Any): Boolean = other.isInstanceOf[Competence]

  override def hashCode(): Int = {
    val state = Seq(getDefinition(), getCompulsory())
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}

File: competence-database/src/main/scala/uzuzjmd/competence/persistence/dao/Competence.java
package uzuzjmd.competence.persistence.dao;

import uzuzjmd.competence.persistence.ontology.Edge;
import uzuzjmd.competence.persistence.ontology.Contexts;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Created by dehne on 11.01.2016.
 */
public class Competence extends AbstractCompetence implements HasDefinition, TreeLike {

    public Boolean compulsory;

    public Competence(String id) {
        super(id);
        compulsory = false;
    }

    public Competence(String id, Boolean compulsory) {
        super(id);
        this.compulsory = compulsory;
    }

    public Competence(String id, LearningProjectTemplate learningProject) {
        super(id);
        this.compulsory = false;
        this.learningProject = learningProject;
    }

    @Override
    public String getDefinition() {
        return this.getId();
    }

    public void addSuggestedCompetenceRequirement(Competence competence) throws Exception {
        createEdgeWith(competence, Edge.SuggestedCompetencePrerequisiteOf);
    }

    public List<Competence> getSuggestedCompetenceRequirements() throws Exception {
        List<Competence> result =  getAssociatedDaosAsDomain(Edge.SuggestedCompetencePrerequisiteOf, Competence.class);
        return result;
    }

    public void addRequiredCompetence(Competence competence) throws Exception {
        deleteEdgeWith(competence, Edge.NotPrerequisiteOf);
        createEdgeWith(competence, Edge.PrerequisiteOf);
    }

    public void addNotRequiredCompetence(Competence competence) throws Exception {
        deleteEdgeWith(competence, Edge.PrerequisiteOf);
        createEdgeWith(competence, Edge.NotPrerequisiteOf);
    }

    public List<Catchword> getCatchwords() throws Exception {
        return getAssociatedDaosAsRange(Edge.CatchwordOf, Catchword.class);
    }

    public List<Competence> getRequiredCompetences() throws Exception {
        return getAssociatedDaosAsRange(Edge.PrerequisiteOf, Competence.class);
    }

    public List<Competence> getSuggestedRequiredCompetences() throws Exception {
        return getAssociatedDaosAsRange(Edge.SuggestedCompetencePrerequisiteOf, Competence.class);
    }

    public String[] getRequiredCompetencesAsArray() throws Exception {
        List<Competence> requiredCompetences = getRequiredCompetences();
        String[] result = new String[requiredCompetences.size()];
        int i = 0;
        for (Competence requiredCompetence : requiredCompetences) {
            result[i] = requiredCompetence.getDefinition();
            i++;
        }
        return result;
    }

    public Boolean isLinkedAsRequired() throws Exception {
        return !(getRequiredCompetences().isEmpty());
    }

    public Boolean isLinkedAsSuggestedRequired() throws Exception {
        return !(getSuggestedCompetenceRequirements().isEmpty());
    }

    public Boolean isAllowed(User user) throws Exception {
        List<Competence> prerequisites = getAssociatedDaosAsDomain(Edge.PrerequisiteOf, Competence.class);
        Boolean result = true;
        for (Competence prerequisite : prerequisites) {
            result = result && prerequisite.hasEdge(user, Edge.UserHasPerformed);
        }
        return result;
    }


    public void addSuperCompetence(Competence superCompetence) throws Exception {
        createEdgeWith(Edge.subClassOf, superCompetence);
    }

    public void removeSuperCompetence(Competence superCompetence) throws Exception {
        deleteEdgeWith(superCompetence, Edge.subClassOf);
    }


    @Override
    public String toString() {
        return this.getId();
    }

    public SelfAssessment getAssessment(User user) throws Exception {
        return queryManager.getSelfAssessment(this, user);
    }

    public List<Competence> getShortestPathToSubCompetence(Competence subCompetence) throws Exception {
        return queryManager.getShortestSubClassPath(subCompetence.getId(), this.getId(), Competence.class);
    }

    public List<Catchword> getCatchwordsAsJava() throws Exception {
        return getCatchwords();
    }

    public void addCatchword(Catchword dao) throws Exception {
        dao.persist();
        dao.createEdgeWith(Edge.CatchwordOf, this);
    }

    public void addLearningTemplate(LearningProjectTemplate learningTemplate) throws Exception {
        createEdgeWith(learningTemplate, Edge.LearningProjectTemplateOf);
    }

    public List<Operator> getOperators() throws Exception {
        return getAssociatedDaosAsDomain(Edge.OperatorOf, Operator.class);
    }

    public void addCourseContext(CourseContext course) throws Exception {
        createEdgeWith(course, Edge.CourseContextOf);
        addSuperCompetencesToCourse(this, course);
    }

    public void addSuperCompetencesToCourse(Competence competence, CourseContext course) throws Exception {
        Set<Competence> superCompetences = competence.listSuperClasses(Competence.class);
        for (Competence superCompetence : superCompetences) {
            superCompetence.createEdgeWith(course, Edge.CourseContextOf);
        }
    }

    public Boolean isSuperClassOf(Competence subCompetence) {
        try {
            return listSubClasses(this.getClass()).contains(subCompetence);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public Boolean isSubClassOf(Competence superCompetence) throws Exception {
        Set<? extends Competence> result = listSuperClasses(this.getClass());
        try {
            return result.contains(superCompetence);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    @Override
    public void deleteTree() throws Exception {
        Set<Competence> subClasses = listSubClasses();
        for (Competence subClass : subClasses) {
            deleteTree();
        }
        this.delete();
    }

    @Override
    public Dao persist() throws Exception {
        super.persist();
        createEdgeWith(Edge.subClassOf, new Competence(DBInitializer.COMPETENCEROOT));
        CourseContext universityContext = new CourseContext(Contexts.university);
        createEdgeWith(universityContext, Edge.CourseContextOf);
        if (learningProject != null) {
            createEdgeWith(learningProject, Edge.LearningProjectTemplateOf);
        }
        return this;
    }


    public Set<Competence> listSubClasses() throws Exception {
        return (HashSet<Competence>) super.listSubClasses(getClass());
    }
}
